providers/fips/self_test_data.c

```diff
@@ -3692,79 +3239,643 @@ static const ST_KAT_PARAM ml_kem_key[] = {
+            ml_dsa_key,
+        },
     },
-# endif
-# if !defined(OPENSSL_NO_SLH_DSA)
+#endif
+#ifndef OPENSSL_NO_SLH_DSA
     {
+        "SLH-DSA-SHA2-128f",
         OSSL_SELF_TEST_DESC_KEYGEN_SLH_DSA,
-        "SLH-DSA-SHA2-128f", 1,
-        slh_dsa_sha2_128f_keygen_init_params,
-        slh_dsa_128f_keygen_expected_params
+        SELF_TEST_KAT_ASYM_KEYGEN,
+        SELF_TEST_DEFERRED,
+        SELF_TEST_STATE_INIT,
+        .u.akgen = {
+            slh_dsa_sha2_128f_keygen_init_params,
+            slh_dsa_128f_keygen_expected_params,
+        },
     },
-# endif
-};
-#endif /* !OPENSSL_NO_ML_DSA || !OPENSSL_NO_SLH_DSA */
-
-static const ST_KAT_ASYM_CIPHER st_kat_asym_cipher_tests[] = {
+#endif
+#ifndef OPENSSL_NO_ML_KEM
     {
+        "ML-KEM-512",
+        OSSL_SELF_TEST_DESC_KEM,
+        SELF_TEST_KAT_KEM,
+        SELF_TEST_DEFERRED,
+        SELF_TEST_STATE_INIT,
+        .u.kem = {
+            ml_kem_key,
+            ITM_BUF(ml_kem_512_cipher_text),
+            ITM_BUF(ml_kem_512_entropy),
+            ITM_BUF(ml_kem_512_secret),
+            ITM_BUF(ml_kem_512_reject_secret),
+        },
+    },
+#endif
+    {
+        "RSA",
         OSSL_SELF_TEST_DESC_ASYM_RSA_ENC,
-        "RSA", 0,
-        1,
-        rsa_pub_key,
-        rsa_enc_params,
-        ITM(rsa_asym_plaintext_encrypt),
-        ITM(rsa_asym_expected_encrypt),
+        SELF_TEST_KAT_ASYM_CIPHER,
+        SELF_TEST_DEFERRED,
+        SELF_TEST_STATE_INIT,
+        ITM_BUF(rsa_asym_plaintext_encrypt),
+        ITM_BUF(rsa_asym_expected_encrypt),
+        .u.ac = {
+            1,
+            rsa_pub_key,
+            rsa_enc_params,
+        },
+        .depends_on = rsaenc_depends_on,
     },
     {
+        "RSA",
         OSSL_SELF_TEST_DESC_ASYM_RSA_DEC,
-        "RSA", 0,
-        0,
-        rsa_priv_key,
-        rsa_enc_params,
-        ITM(rsa_asym_expected_encrypt),
-        ITM(rsa_asym_plaintext_encrypt),
+        SELF_TEST_KAT_ASYM_CIPHER,
+        SELF_TEST_DEFERRED,
+        SELF_TEST_STATE_INIT,
+        ITM_BUF(rsa_asym_expected_encrypt),
+        ITM_BUF(rsa_asym_plaintext_encrypt),
+        .u.ac = {
+            0,
+            rsa_priv_key,
+            rsa_enc_params,
+        },
+        .depends_on = rsaenc_depends_on,
     },
     {
+        "RSA",
         OSSL_SELF_TEST_DESC_ASYM_RSA_DEC,
-        "RSA", 0,
-        0,
-        rsa_crt_key,
-        rsa_enc_params,
-        ITM(rsa_asym_expected_encrypt),
-        ITM(rsa_asym_plaintext_encrypt),
+        SELF_TEST_KAT_ASYM_CIPHER,
+        SELF_TEST_DEFERRED,
+        SELF_TEST_STATE_INIT,
+        ITM_BUF(rsa_asym_expected_encrypt),
+        ITM_BUF(rsa_asym_plaintext_encrypt),
+        .u.ac = {
+            0,
+            rsa_crt_key,
+            rsa_enc_params,
+        },
+        .depends_on = rsaenc_depends_on,
+    },
+    {
+        "HMAC",
+        OSSL_SELF_TEST_DESC_INTEGRITY_HMAC,
+        SELF_TEST_KAT_MAC,
+        SELF_TEST_DEFERRED,
```

Comment:
HMAC is not deferred

Reply:
You probably need a separate state for this one, since it is not onload either, its a integrity phase check only.

Reply:
When we do 'on demand' tests i would like it to just run the tests it was running before you added deferred testing..
i.e. NOT SHA2 for example. The same holds if the user decides not to run in deferred mode.